internal/state/registry.go

```diff
@@ -2,68 +2,94 @@ package state
+			v1alpha1.ConditionReasonConfigurationErr,
+			err,
+		)
+		return nextState(sFnUpdateStatusWithError(err))
 	}
-	// overwrite registryStatus when there is serverless (not serverless-manager) cluster-wide external registry secret
-	setExternalRegistrySecretNameInDockerRegistryStatus(ctx, r, s)
 
+	// update status if needed
 	if s.snapshot.DockerRegistry != s.instance.Status.DockerRegistry {
 		return nextState(sFnUpdateStatusAndRequeue)
 	}
 
 	return nextState(sFnOptionalDependencies)
 }
 
-func setExternalRegistrySecretNameInDockerRegistryStatus(ctx context.Context, r *reconciler, s *systemState) {
-	// doc: https://kyma-project.io/docs/kyma/latest/05-technical-reference/svls-03-switching-registries#cluster-wide-external-registry
+func configureRegistry(ctx context.Context, r *reconciler, s *systemState) error {
 	secret, err := registry.GetServerlessExternalRegistrySecret(ctx, r.client, s.instance.GetNamespace())
-	// ignore errors because it only set status
-	if err != nil || secret == nil {
-		return
+	if err != nil {
+		return err
 	}
-	if address, ok := secret.Data["serverAddress"]; ok {
-		s.instance.Status.DockerRegistry = string(address)
+
+	switch {
+	case secret != nil:
+		// case: use runtime secret (with labels)
+		// doc: https://kyma-project.io/docs/kyma/latest/05-technical-reference/svls-03-switching-registries#cluster-wide-external-registry
+		setRuntimeRegistryConfig(secret, s)
+	case s.instance.Spec.DockerRegistry.SecretName != nil:
+		// case: use secret from secretName field
+		err := setExternalRegistryConfig(ctx, r, s)
+		if err != nil {
+			return err
+		}
+	case *s.instance.Spec.DockerRegistry.EnableInternal:
+		// case: use internal registry
+		err := setInternalRegistryConfig(ctx, r, s)
+		if err != nil {
+			return err
+		}
+	default:
+		// case: use k3d registry
+		setK3dRegistryConfig(s)
+	}
+
+	addRegistryConfigurationWarnings(secret, s)
+	return nil
+}
+
+func addRegistryConfigurationWarnings(secret *corev1.Secret, s *systemState) {
+	// runtime secret exist and it's other than this under secretName
+	if secret != nil &&
+		s.instance.Spec.DockerRegistry.SecretName != nil &&
+		secret.Name != *s.instance.Spec.DockerRegistry.SecretName {
+		s.addWarning(fmt.Sprintf("actual registry configuration comes from %s/%s and it's different from spec.dockerRegistry.secretName. Reflect the %s secret in the secretName field or delete it", secret.Name, secret.Namespace, secret.Name))
 	}
-	return
+
+	// runtime secret exist and secretName field is empty
+	if secret != nil && s.instance.Spec.DockerRegistry.SecretName == nil {
+		s.addWarning(fmt.Sprintf("actual registry configuration comes from %s/%s and it's different from spec.dockerRegistry.secretName. Reflect %s secret in the secretName field", secret.Name, secret.Namespace, secret.Name))
+	}
+
+	// enableInternal is true and secretName is used
+	if *s.instance.Spec.DockerRegistry.EnableInternal && s.instance.Spec.DockerRegistry.SecretName != nil {
```

Comment:
What should happen when as I user I configure namespaced secret for registry in `kyma-system`?
https://kyma-project.io/docs/kyma/latest/05-technical-reference/svls-03-switching-registries#namespace-scoped-external-registry

Reply:
nothing. Namespace-scoped configuration should stay as it is